### PR TITLE
OEP-7: Python 3 migration plan

### PR DESCRIPTION
**Pull request by jcdyer**
Plan for migrating Open edX code to Python 3.

### CODE REVIEW

File: oeps/oep-0007.rst [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
++---------------+-------------------------------------------+
+
+
+Abstract
+========
+
+With primary development of Python occurring on Python 3, and with Python 2
+scheduled for end-of-life in 2020, edX needs to plan for a transition of all of
+our code to Python 3.  Open edX is a large project, spanning many applications
+over even more github repos, with even more dependencies on third party
```
Comment (clintonb) [inline oeps/oep-0007.rst]: s/github repos/repositories

File: oeps/oep-0007.rst [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+libraries.  This document outlines how we plan to scope the problem, socialize
+understanding of the differences between versions of python, migrate our code,
```
Comment (clintonb) [inline oeps/oep-0007.rst]: Capitalize Python

File: oeps/oep-0007.rst [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+and work with our Open edX community to ensure that the process is as painless
+as possible, and meets the needs of our stakeholders.
+
+
+Rationale
+=========
+
+    * Python 2 will only receive support from the python core devs until 2020.
```
Comment (clintonb) [inline oeps/oep-0007.rst]: Capitalize Python.

File: oeps/oep-0007.rst [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+
+If you find other incompatibilities, a shim will likely be found in [the `six`
+library](https://pythonhosted.org/six/).  For incompatibilies with no other
+solution, edX will maintain a repository of compatibility shims (edx-compat?).
+Ideally, all edx-maintained code that implements different behavior based on
+python version will be in this repo.
+
+
+Choosing a python version
++++++++++++++++++++++++++
```
Comment (jcdyer) [inline oeps/oep-0007.rst]: Why polyglot?

File: oeps/oep-0007.rst [hunk 5]
```diff
@@ -0,0 +1,354 @@
+===============================
+OEP-0007: Migrating to Python 3
+===============================
++---------------+-------------------------------------------+
+| OEP           | :doc:`OEP-7 </oeps/oep-0007>`             |
```
Comment (nedbat) [inline oeps/oep-0007.rst]: When you rebase onto master, you'll need to change this to `<oep-0007>`.

File: oeps/oep-0007.rst [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+library](https://pythonhosted.org/six/).  For incompatibilies with no other
+solution, edX will maintain a repository of compatibility shims (edx-compat?).
+Ideally, all edx-maintained code that implements different behavior based on
+python version will be in this repo.
+
+
+Choosing a python version
++++++++++++++++++++++++++
+
+All greenfield python projects should be written using python 3 unless a
```
Comment (jcdyer) [inline oeps/oep-0007.rst]: s/greenfield/new/

File: oeps/oep-0007.rst [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+Other problems
+--------------
+
+If you find other incompatibilities, a shim will likely be found in [the `six`
+library](https://pythonhosted.org/six/).  For incompatibilies with no other
+solution, edX will maintain a repository of compatibility shims (edx-compat?).
+Ideally, all edx-maintained code that implements different behavior based on
+python version will be in this repo.
+
```
Comment (jcdyer) [inline oeps/oep-0007.rst]: Packaging: Use proper trove classifiers.

File: oeps/oep-0007.rst [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+Which path is best may depend on the enthusiasm of the maintainers for
+supporting Python 3, the amount of resources we want to commit to the project,
+and the availability and quality of alternatives.
+
+__future__ imports
+------------------
+
+All files should have the main `__future__` imports at the top to regularize
+some behaviors that differ by default between Python 2 and 3. `from
+__future__ import absolute_import` forbid the use of implicit relative
```
Comment (maxrothman) [inline oeps/oep-0007.rst]: Nit: these would be more readable as a list

File: oeps/oep-0007.rst [hunk 9]
```diff
@@ -0,0 +1,462 @@
...
+should only support version 2.7.  Future version deprecations are outside the
+scope of this document.
+
+The end goal is for all of our services to be deployed under Python 3 or
+replaced by newer services before the April 2020 end of support.
+
+
+Migrating code
+++++++++++++++
+
```
Comment (maxrothman) [inline oeps/oep-0007.rst]: We should have links to six and future's documentation somewhere in this document

File: oeps/oep-0007.rst [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+There are two major ways of handling text and byte literals uniformly across
+Python versions.  We do not explicitly require one way over the other, but
+decisions should be made on a per-project basis, and adhered to by all
+developers working on that project.
+
+Handling literals, Option 1: Python 3-Style
+...........................................
+
+In order to write code that looks as much like native Python 3 as possible,
+you may want to use `from __future__ import unicode_literals`, which makes bare
```
Comment (maxrothman) [inline oeps/oep-0007.rst]: \`this` does not make code formatting, it makes quote formatting

File: oeps/oep-0007.rst [hunk 11]
```diff
@@ -0,0 +1,462 @@
...
+--------------------------------
+
+Instead of using `d.iterkeys()`, use `future.utils.viewkeys(d)` or
+`six.viewkeys(d)`.  If you need a list, use `list(*.viewkeys(d))`.
+Other similar functions exist for `itervalues()` and `iteritems()`.  These
+changes cannot be made cleanly in the import headers, and will require more
+work to change after the fact.  This can be avoided in some cases by iterating
+directly over the dict object.  Instead of using:
+
+    for key, value in six.viewitems(d):
```
Comment (maxrothman) [inline oeps/oep-0007.rst]: These format as quote blocks, not code blocks

File: oeps/oep-0007.rst [hunk 12]
```diff
@@ -0,0 +1,462 @@
...
+
+Ideally, all listed minor versions should be tested in a continuous integration
+environment.  At a minimum, at least the lowest and highest minor versions of
+each supported major version must be tested.
+
+Other problems
+--------------
+
+If you find other incompatibilities, a shim will likely be found in [the `six`
+library](https://pythonhosted.org/six/).  For incompatibilies with no other
```
Comment (maxrothman) [inline oeps/oep-0007.rst]: Ah, here's the link. It probably should've been up at the first reference.

File: oeps/oep-0007.rst [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+we need to update the official edX code style guide to mandate compatible code
+practices.  To socialize these practices among our engineers, we will announce
+our efforts to migrate to Python 3 during an engineering all-hands meeting,
+offer a workshop in writing compatible code, and promote awareness of
+incompatibilities during code reviews.  Additionally, we will host regular
+Python 3 office hours to help answer questions and troubleshoot problems that
+arise during migration.
+
+Appropriate tooling will help.  Tests should be configured to run under both
+Python 2 or 3 (tox helps with this).  As it would double the cost of running
```
Comment (maxrothman) [inline oeps/oep-0007.rst]: link to tox

File: oeps/oep-0007.rst [hunk 14]
```diff
@@ -0,0 +1,462 @@
...
+the explictness of the compatibility code will make it easier to find
+compatibility issues, and to avoid introducing new issues.
+
+Third-party dependencies
+------------------------
+
+Before converting a codebase to Python 3, we need to make sure the code we
+depend on will also support Python 3.  We can make a rough check for
+third-party library support using the [Can I Use Python 3 website or
+library](https://caniusepython3.com/).  This will sometimes show inaccurate
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: This link and 5 others below use Markdown formatting, but this is a reStructuredText document.  They need to be updated to render correctly in the output.

File: oeps/oep-0007.rst [hunk 15]
```diff
@@ -0,0 +1,462 @@
...
+double-slash division (`a // b`) always truncate down.
+
+Text handling
+-------------
+
+Text handling is the largest area of difficulty in porting Python.  Where
+possible, we will use unambiguous text or byte objects.  In most cases, text
+should be preferred.  Bytes should only be used when you can answer the
+question: "Do I need this specific sequence of bytes."  The most
+error-resistent way to acheive this is to use what is called a "unicode
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: Spelling: resistent -> resistant

File: oeps/oep-0007.rst [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+sandwich."  This means that as soon as you receive data from a file or network
+interface, it should be converted to text. Your code should then treat it as
+text for as long as possible, only encoding it back to bytes when sending it to
+an interface that requires bytes (such as a file, a network interface, or a
+bytes-oriented library). The only operation that should (ideally) be performed
+on bytes is decoding.
+
+In those cases where ambiguity is required (such as working with libraries like
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: The matching close parenthesis is missing

File: oeps/oep-0007.rst [hunk 17]
```diff
@@ -0,0 +1,462 @@
...
+an interface that doesn't specify its encoding, pass it through a wrapper that
+does specify the encoding, and use that wrapper instead.
+
+There are two major ways of handling text and byte literals uniformly across
+Python versions.  We do not explicitly require one way over the other, but
+decisions should be made on a per-project basis, and adhered to by all
+developers working on that project.
+
+Handling literals, Option 1: Python 3-Style
+...........................................
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: Periods don't seem to be among the [valid section header underline characters](http://docutils.sourceforge.net/docs/user/rst/quickref.html#section-structure)

File: oeps/oep-0007.rst [hunk 18]
```diff
@@ -0,0 +1,462 @@
...
+unicode_literals`, but instead recommend using explicit `u'unicode'` and
+`b'byte'` literals throughout. Bare native-string literals should be used
+sparingly, and explicitly called out as intentional usages.  This "calling out"
+can be enforced by installing the `caniusepython3` pylint extension, which will
+flag a warning (`native-string`) on such uses.  A native string would then be
+instantiated as:
+
+    native = 'string'  # pylint: disable=native-string
+
+This version creates noisier code than Option 2, below, but makes it easier to
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: This is Option 2; did you mean to reference Option 1 above?

File: oeps/oep-0007.rst [hunk 19]
```diff
@@ -0,0 +1,462 @@
...
+
+All packages should maintain the proper trove classifiers for the versions of
+Python they support.
+
+In the following recommendations, the major version classifiers comprise:
+
+    Programming Language :: Python :: 2
+    Programming Language :: Python :: 2 :: Only
+    Programming Language :: Python :: 3
+    Programming Language :: Python :: 3 :: Only
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: The odd-numbered lines from these blocks are being omitted from the output; the colons are probably throwing off the formatting somehow

File: oeps/oep-0007.rst [hunk 20]
```diff
@@ -0,0 +1,462 @@
...
+our efforts to migrate to Python 3 during an engineering all-hands meeting,
+offer a workshop in writing compatible code, and promote awareness of
+incompatibilities during code reviews.  Additionally, we will host regular
+Python 3 office hours to help answer questions and troubleshoot problems that
+arise during migration.
+
+Appropriate tooling will help.  Tests should be configured to run under both
+Python 2 or 3 (tox helps with this).  As it would double the cost of running
+tests to run both all the time, we may only want to enable Python 3 tests in
+the CI environment for repos that are actively being migrated.  We still want
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: The extra test runs shouldn't matter for repos that use Travis, as they're run in parallel on external hardware and don't incur usage fees.

File: oeps/oep-0007.rst [hunk 21]
```diff
@@ -0,0 +1,462 @@
...
+* IDAs that we want to continue supporting in the future
+  * Old IDAs (that we want to replace) should not be upgraded, but we will need
+    to prioritize replacement to occur during the migration timeframe.
+* Implement remote execution of xBlocks (to allow a window of bicompatibility
+  for external xblocks)
+* edx-platform
+  * Deploy xblocks separately to test remote execution
+  * Migrate to Python 3
+  * Upgrade external xblocks as needed, and support partners who wish to do the
+    same
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: This isn't valid rst formatting for nested lists; the type of bullet needs to change for the 2nd level

File: oeps/oep-0007.rst [hunk 22]
```diff
@@ -0,0 +1,462 @@
...
+Code conversion should be automated as much as possible.  The `future` library
+includes a `futurize` script that will do much of the legwork.  As we gain
+experience migrating code, we will develop a sense as to how aggressively we
+can use `futurize`, and what other work needs to be done.
+
+
+Deploying Services
+++++++++++++++++++
+
+Once a service has been achieved full Python 3 compatibility, we will need to
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: Grammar, probably just need to drop "been"

File: oeps/oep-0007.rst (hunk 20 above)
Comment (maxrothman) [inline oeps/oep-0007.rst]: They will eat into our available travis workers, which we're already a little short on, leading to more queue backup and longer delays for tests to run.

File: oeps/oep-0007.rst [hunk 23]
```diff
@@ -0,0 +1,462 @@
...
+Migrating code
+++++++++++++++
+
+All new code in Python 2 codebases should be written to be compatible with Python 3.  There
+are two major libraries to help work with compatibility differences.  Both have
+their uses, so rather than mandating one over the other, we offer situations
+where each one is preferable.  The two options are `six`, and `future`.
+
+Six is a simple compatibility library.  If something works differently in
+Python 2 and and Python 3, you import six, and use functions that it provides
```
Comment (bmedx) [inline oeps/oep-0007.rst]: "and and"

Comment (jcdyer) [inline oeps/oep-0007.rst]: fixed.

File: oeps/oep-0007.rst (hunk 19 above)
Comment (jcdyer) [inline oeps/oep-0007.rst]: Shows up now.  I'm not sure if I changed something or not.

**Comment by jcdyer** [comment]
@jmbowman: I think I've responded to all your concerns.  Please take another look.

File: oeps/oep-0007.rst [hunk 24]
```diff
@@ -0,0 +1,523 @@
...
+6. Deploy the project in Python 3.
+7. Stop testing under Python 2.
+
+
+Order of migrations
++++++++++++++++++++
+
+* IDAs that we want to continue supporting in the future
+  + Old IDAs (that we want to replace) should not be upgraded, but we will need
+    to prioritize replacement to occur during the migration timeframe.
```
Comment (jmbowman) [inline oeps/oep-0007.rst]: The nested bullets here and in the third bullet point need to be [separated by blank lines](http://www.sphinx-doc.org/en/1.5.1/rest.html#lists-and-quote-like-blocks) from the parent list items to be formatted correctly

File: oeps/oep-0007.rst [hunk 25]
```diff
@@ -0,0 +1,526 @@
...
+our efforts to migrate to Python 3 during an engineering all-hands meeting,
+offer a workshop in writing compatible code, and promote awareness of
+incompatibilities during code reviews.  Additionally, we will host regular
+Python 3 office hours to help answer questions and troubleshoot problems that
+arise during migration.
+
+Appropriate tooling will help.  Tests should be configured to run under both
+Python 2 or 3 using tox.  As it would double the cost of running
+tests to run both all the time, we may only want to enable Python 3 tests in
+the CI environment for repos that are actively being migrated.  We still want
```
Comment (benpatterson) [inline oeps/oep-0007.rst]: Please don't worry about this. We are building our CI infrastructure for scale. If we reach a point where we can't support what we need to do, then we can think about removing pieces...

Comment (benpatterson) [inline oeps/oep-0007.rst]: Also, technically...due to our billing, prices won't be doubled by using more builds in Travis, for example. So the statement is inaccurate. I argue that you remove this value statement altogether; we should just test what we need to test.

File: oeps/oep-0007.rst [hunk 26]
```diff
@@ -0,0 +1,526 @@
...
+Choosing a Python version
++++++++++++++++++++++++++
+
+All new Python projects should be written using Python 3 unless a compelling
+reason (such as incompatible support libraries with no reasonable alternative
+available) necessitates that we stick with Python 2.
+
+
+New libraries should additionally support Python 2 using a single code-base
+with Six_ as a compatibility layer if they will be used by projects that run on
```
Comment (mulby) [inline oeps/oep-0007.rst]: We have, at times, struggled with different libraries requiring different versions of `six`.

File: oeps/oep-0007.rst [hunk 27]
```diff
@@ -0,0 +1,526 @@
...
+support Python 3 that we can perform, we should opt to push those changes
+upstream rather than forking projects when possible.
+
+Code conversion should be automated as much as possible.  The future_ library
+includes a ``futurize`` executable that will do much of the legwork.  As we gain
+experience migrating code, we will develop a sense as to how aggressively we
+can use ``futurize``, and what other work needs to be done.
+
+
+Deploying Services
```
Comment (mulby) [inline oeps/oep-0007.rst]: I think you could remove this section. This is true of all significant changes like this and only applicable to edx.org

File: oeps/oep-0007.rst [hunk 28]
```diff
@@ -0,0 +1,526 @@
...
++---------------+-------------------------------------------+
+| OEP           | :doc:`OEP-7 <oep-0007>`                   |
++---------------+-------------------------------------------+
+| Title         | Migrating to Python 3                     |
++---------------+-------------------------------------------+
+| Last-Modified | 2016-09-23                                |
++---------------+-------------------------------------------+
+| Author        | Cliff Dyer <cdyer@edx.org>                |
++---------------+-------------------------------------------+
+| Arbiter       |                                           |
```
Comment (mulby) [inline oeps/oep-0007.rst]: Don't forget to update this with the actual arbiter (not sure who it is for this OEP).

File: oeps/oep-0007.rst [hunk 29]
```diff
@@ -0,0 +1,526 @@
...
+All files should have the main ``__future__`` imports at the top to regularize
+some behaviors that differ by default between Python 2 and 3.
+
+* ``from __future__ import absolute_import`` prevents the use of implicit
+  relative imports
+* ``from __future__ import print_function`` makes print a function instead of a
+  statement.
+* ``from __future__ import division`` will make single-slash division
+  (``a / b``) always perform floating point division, and double-slash division
+  (``a // b``).
```
Comment (blarghmatey) [inline oeps/oep-0007.rst]: Incomplete sentence. Should read `and double slash division ("a // b") perform integer division`

**Comment by pdpinch** [comment]
Since course content can contain python code, I think you need a section that addresses it specifically. 

I presume the python version is controlled by the sandbox setting. But I think there's a feature flag for disabling the sandbox, in which case I think python in courseware would use the same version that edx-platform is running on.

**Comment by jcdyer** [comment]
Great idea @pdpinch.  I'll add that in.

File: oeps/oep-0007.rst (hunk 25 above)
Comment (jcdyer) [inline oeps/oep-0007.rst]: Thanks, Ben.  That's helpful information.

File: oeps/oep-0007.rst (hunk 28 above)
Comment (jcdyer) [inline oeps/oep-0007.rst]: Done.

File: oeps/oep-0007.rst (hunk 29 above)
Comment (jcdyer) [inline oeps/oep-0007.rst]: Fixed.

**Comment by pdpinch** [comment]
credit should go to @Colin-Fredericks for calling it out -- I'm just a (very) interested party

**Comment by jcdyer** [comment]
Thanks @Colin-Fredericks!  I've added language to address that.  (Also, I think we went to high school together.)

**Comment by Colin-Fredericks** [comment]
@jcdyer Huh! So we did! Now I need to dig out my old 1996 NMH Facebook and find your picture. Thanks for the help here.

**Comment by jcdyer** [comment]
@Colin-Fredericks Think big hair.

@jmbowman I believe this OEP is ready to merge.

**Pull request merged**